# Numerical Solution of Equations

In many scientific and engineering problems, we need to find the value of a variable that satisfies an equation of the form

$$
f(x) = 0
$$

Such problems are known as **root-finding problems**, because the goal is to find the **root** (or zero) of the function \(f(x)\). A root is a value \(x = r\) for which the function becomes zero:

$$
f(r) = 0
$$

For some simple equations, it is possible to find the exact solution using algebraic methods. However, in most real-world applications the function may be complicated, nonlinear, or even impossible to solve analytically. In these cases, we use **numerical methods** to approximate the root.

Numerical root-finding methods use iterative procedures to gradually approach the solution. Starting from an initial guess or an interval that contains the root, these algorithms generate better approximations until the desired accuracy is achieved.

Several numerical techniques exist for solving nonlinear equations, Each method has its own advantages, convergence properties, and computational cost. In this chapter, we will explore some of these methods and implement them using **Python** to better understand how they work in practice.

The first method we will study is the **Bisection Method**, which is one of the simplest and most reliable root-finding algorithms.


## Determining Roots with Desired Accuracy

Suppose we want to solve the nonlinear equation

$$
f(x) = 0
$$

In numerical methods, instead of computing the exact root, we usually construct a **sequence of approximations** $x_n$ that gradually approaches the true root $\alpha$.

Mathematically, this idea can be expressed as

$$
\lim_{n \to \infty} x_n = \alpha
$$

This means that as the number of iterations increases, the approximation $x_n$ becomes closer to the true root $\alpha$:

$$
x_n \approx \alpha
$$

---

## Locating the Root

In practice, to determine a root of an equation with a desired accuracy, we first try to find a **small interval that contains the root**.

Two important conditions are usually considered.

### Condition A: Existence of a Root in the Interval

There must exist an interval $[a,b]$ that contains at least one root.

Mathematically:

1. The function $f(x)$ must be **continuous** on the interval $[a,b]$.

2. The function values at the endpoints must have **opposite signs**:

$$
f(a)f(b) < 0
$$

According to the **Intermediate Value Theorem**, if these conditions hold, then there exists at least one number $\alpha \in [a,b]$ such that

$$
f(\alpha) = 0
$$

---

### Condition B: Uniqueness of the Root

To ensure that the root is **unique** in the interval, an additional condition can be imposed:

$$
f'(x) \ne 0 \qquad \text{for all } x \in [a,b]
$$

This guarantees that the function is monotonic in the interval, and therefore the root in $[a,b]$ is unique.

---

## Stopping Criteria in Numerical Methods

When implementing iterative numerical algorithms, we must decide **when to stop the computations**. This is determined by a **stopping criterion**.

Common stopping criteria include:

### 1. Function Value Tolerance

If the function value becomes sufficiently small,

$$
|f(x_n)| < \varepsilon
$$

where $\varepsilon$ is a predefined tolerance (for example $10^{-6}$), then the current approximation $x_n$ is accepted as the root.

---

### 2. Successive Approximation Difference

If the difference between two successive approximations becomes very small,

$$
|x_n - x_{n-1}| < \varepsilon
$$

then the algorithm stops and $x_n$ is taken as the approximate root.

---

### 3. Maximum Number of Iterations

Sometimes a maximum number of iterations $N$ is specified.  
If the number of iterations exceeds this limit, the algorithm stops and the last approximation is accepted as the result.


## Bisection Method

The **Bisection Method** is one of the simplest and most reliable numerical techniques for finding a root of the equation

$$
f(x) = 0
$$

This method is based on the **Intermediate Value Theorem**.  
If a function $f(x)$ is continuous on an interval $[a,b]$ and

$$
f(a)f(b) < 0
$$

then there exists at least one root $\alpha$ in the interval $(a,b)$.

The main idea of the bisection method is to **repeatedly divide the interval into two halves** and select the subinterval that contains the root.

---

## Idea of the Method

Starting with an interval $[a,b]$ such that

$$
f(a)f(b) < 0
$$

we compute the midpoint

$$
x = \frac{a + b}{2}
$$

Then we evaluate the function at this midpoint and determine which half of the interval contains the root.

This process is repeated until the root is found with the desired accuracy.

---

## Bisection Algorithm

Given an initial interval $[a,b]$ where $f(a)f(b) < 0$:

### Step 1

Compute the midpoint

$$
x = \frac{a + b}{2}
$$

---

### Step 2

If

$$
f(a)f(x) < 0
$$

then the root lies in the interval

$$
(a,x)
$$

so we update

$$
b = x
$$

and repeat the process with the new interval $[a,b]$.

---

### Step 3

If

$$
f(a)f(x) > 0
$$

then the root lies in the interval

$$
(x,b)
$$

so we update

$$
a = x
$$

and repeat the process with the new interval $[a,b]$.

---

### Step 4

If

$$
f(a)f(x) = 0
$$

then

$$
x
$$

is exactly the root and the algorithm terminates.

---

## Stopping Condition

In practice, the equality

$$
f(a)f(x) = 0
$$

rarely occurs due to numerical computations. Therefore, we usually stop the algorithm when one of the following conditions is satisfied:

$$
|f(x)| < \varepsilon
$$

where $\varepsilon$ is the desired accuracy.

---

## Important Property

The **Bisection Method always converges** provided that

- $f(x)$ is continuous on $[a,b]$
- $f(a)f(b) < 0$

Therefore, the root is **guaranteed to be found with the desired accuracy**.


## Problem

Approximate a root of the nonlinear equation

$$
x^3 + x - 1 = 0
$$

using the **Bisection Method** with an accuracy

$$
\varepsilon < 10^{-3}
$$


### Interval Selection for the Bisection Method

Before applying the Bisection Method, we must first find an interval $[a,b]$ that contains a root of the equation.

This condition is based on the **Intermediate Value Theorem**.  
If a function $f(x)$ is continuous on the interval $[a,b]$ and the values of the function at the endpoints have **opposite signs**, then there exists at least one root in that interval.

Mathematically, this condition is

$$
f(a) \cdot f(b) < 0
$$

This inequality indicates that the function changes sign between $a$ and $b$, meaning the graph of the function crosses the x-axis somewhere inside the interval.

Therefore, before starting the Bisection iterations, we check whether the selected interval satisfies this condition.  
If the condition is not satisfied, a new interval must be chosen.


In [41]:
import numpy as np
from tabulate import tabulate

In [42]:
def f(x):
    return x**3 + x - 1

In [ ]:
try:
    a, b = [float(i) for i in input("Enter the interval endpoints a,b (separated by comma): ").split(',')]

    while True:

        f_a = f(a)
        f_b = f(b)

        print(f"f(a) = {f_a} , f(b) = {f_b}")

        if f_a * f_b < 0:
            print("Valid interval ✅")
            print("Since f(a) * f(b) < 0, the interval contains at least one root.")
            break
        else:
            print("Invalid interval ❌")
            print("f(a) and f(b) have the same sign, so the interval may not contain a root.")
            print("Please enter a new interval.\n")

            a, b = [int(i) for i in input("Enter the interval endpoints a,b (separated by comma): ").split(',')]

except Exception as e:
    print(f"An error occurred: {e}")


f(a) = -1 , f(b) = 1
Valid interval ✅
Since f(a) * f(b) < 0, the interval contains at least one root.


### Root Uniqueness Condition

After finding an interval that contains a root, it is useful to check whether the root in that interval is **unique**.

A sufficient condition for uniqueness is that the derivative of the function does not change sign in the interval $[a,b]$ and is never zero:

$$
f'(x) \neq 0 \quad \text{for all } x \in [a,b]
$$

If the derivative keeps a constant sign (always positive or always negative), the function is **monotonic** on that interval.

- If $f'(x) > 0$ → the function is strictly increasing.  
- If $f'(x) < 0$ → the function is strictly decreasing.

A monotonic function can cross the x-axis **at most once**, which guarantees that the root in the interval is **unique**.


In [ ]:
def df(x):
    return 3*(x**2) + 1

In [ ]:
xs = np.linspace(a, b, 500)

if all(round(df(x), 5) != 0 for x in xs):
    print("The root in this interval is unique.")
else:
    print("The root may not be unique.")

The root in this interval is unique.


### Bisection Iteration Process

After selecting a valid interval that contains a root, the Bisection Method proceeds by repeatedly cutting the interval in half.

In each iteration:

1. The midpoint is computed:
   $$
   c = \frac{a + b}{2}
   $$

2. We evaluate the function at the endpoints and at the midpoint:  
   $f(a)$, $f(b)$, and $f(c)$.

3. Using the sign of $f(c)$, we determine which half of the interval contains the root:
   - If $f(a)f(c) < 0$, the root lies in the left half → set $b = c$
   - Otherwise, the root lies in the right half → set $a = c$

4. This process progressively narrows the interval around the root.

5. The algorithm stops when the midpoint becomes sufficiently accurate, determined by:
   $$
   |f(c)| < \varepsilon
   $$

At each iteration, the interval becomes smaller, guaranteeing convergence to the true root because the Bisection Method is based on the Intermediate Value Theorem and always preserves a root inside the updated interval.


In [47]:
iteration  = 1
epsilon = 1e-3

rows = []

while True:
    c = (a + b) / 2

    f_a = f(a)
    f_b = f(b)
    f_c = f(c)

    rows.append([iteration, a, b, c, f_a, f_c])

    if f_a * f_c < 0:
        b = c
    else :
        a = c
    
    if np.abs(f_c) < epsilon:
        break
    
    iteration  += 1

headers = ["iteration", "a", "b", "c = (a + b) / 2", "f(a)", "f(c)"]
print(tabulate(rows, headers=headers, floatfmt="0.6f"))


  iteration         a         b    c = (a + b) / 2       f(a)       f(c)
-----------  --------  --------  -----------------  ---------  ---------
          1  0.000000  1.000000           0.500000  -1.000000  -0.375000
          2  0.500000  1.000000           0.750000  -0.375000   0.171875
          3  0.500000  0.750000           0.625000  -0.375000  -0.130859
          4  0.625000  0.750000           0.687500  -0.130859   0.012451
          5  0.625000  0.687500           0.656250  -0.130859  -0.061127
          6  0.656250  0.687500           0.671875  -0.061127  -0.024830
          7  0.671875  0.687500           0.679688  -0.024830  -0.006314
          8  0.679688  0.687500           0.683594  -0.006314   0.003037
          9  0.679688  0.683594           0.681641  -0.006314  -0.001646
         10  0.681641  0.683594           0.682617  -0.001646   0.000694


### Result and Interpretation

From the table of iterations, the interval containing the root is gradually reduced at each step.  
Starting from the initial interval $[0,1]$, the midpoint of the interval is computed and the subinterval containing the root is selected according to the sign of the function.

After several iterations, the algorithm stops when the stopping criterion

$$
|f(c)| < 10^{-3}
$$

is satisfied.

From the last iteration in the table:

$$
c \approx 0.682617
$$

and

$$
f(c) \approx 0.000694
$$

Since $|f(c)| < 10^{-3}$, the condition is satisfied and the algorithm terminates.  
Therefore, the approximate root of the equation

$$
x^3 + x - 1 = 0
$$

is

$$
x \approx 0.6826
$$

with an accuracy better than $10^{-3}$.

---

### Advantages of the Bisection Method

- **Guaranteed convergence:** If $f(a)f(b) < 0$, the method will always converge to a root.
- **Simplicity:** The algorithm is very easy to understand and implement.
- **Stability:** It is numerically stable and does not diverge.

---

### Disadvantages of the Bisection Method

- **Slow convergence:** The method converges linearly, so it usually requires many iterations.
- **Requires a valid interval:** An initial interval where the function changes sign must be known.
- **Not efficient for high precision:** Compared to methods such as Newton's Method, it reaches high accuracy more slowly.
